cpp/src/parquet/encoding_test.cc

```diff
@@ -540,8 +538,203 @@ void GetBinaryDictDecoder(DictEncoder<ByteArrayType>* encoder, int64_t num_value
 
   *out_values = buf;
   *out_dict = dict_buf;
-  *out_decoder = std::unique_ptr<ByteArrayDecoder>(
-      dynamic_cast<ByteArrayDecoder*>(decoder.release()));
+  ASSERT_NE(decoder, nullptr);
+  auto released = dynamic_cast<TypedDecoder<T>*>(decoder.release());
+  ASSERT_NE(released, nullptr);
+  *out_decoder = std::unique_ptr<TypedDecoder<T>>(released);
+}
+
+template <typename Pair = void>
+class EncodingAdHocTyped {
+ public:
+  using Types = ::testing::Types<std::pair<::arrow::BooleanType, BooleanType>,
+                                 std::pair<::arrow::Int32Type, Int32Type>,
+                                 std::pair<::arrow::Int64Type, Int64Type>,
+                                 std::pair<::arrow::FloatType, FloatType>,
+                                 std::pair<::arrow::DoubleType, DoubleType>,
+                                 std::pair<::arrow::FixedSizeBinaryType, FLBAType>>;
+};
+
+template <typename ArrowType, typename ParquetType>
+class EncodingAdHocTyped<std::pair<ArrowType, ParquetType>> : public ::testing::Test {
```

Comment:
What's the benefit of using the pair value as a template parameter? Wasn't obvious from cursory glance

Reply:
For each test, I needed the ParquetType and the corresponding ArrowType (for usage of RandomArrayGenerator, TypeTraits::type_singleton). I could also have accomplished this by defining EncodingTraits<ParquetType>::ArrowType or so if you prefer.

Reply:
I went ahead and refactored to do this